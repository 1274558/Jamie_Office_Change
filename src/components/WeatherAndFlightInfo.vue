<template>
  <v-container>
    <v-row>
      <v-col>
        <v-text-field
            v-model="sourcecity"
            label="Where are you flying from? (Enter airport code)"
            required
        ></v-text-field>
        <p v-if="errors">
          <label>{{errorLabel}}</label>
          <label>{{ errors }}</label>
        </p>
        <v-btn 
          type="submit" 
          color="warning" 
          @click="fetchFlightDetails()" 
          dark
          >
          {{submitButtonText}}
          </v-btn>
      </v-col>
    </v-row>
    <v-row>
      <v-col 
        cols="4" 
        v-for="item in cities" 
        v-bind:key="item.city"
        >
        <city-info-card 
          :cityData="item" 
          v-on:choose-city="chooseTheCity"/>
      </v-col>
    </v-row>
    <div>
      <v-alert 
        type="success"
        v-model="alert"
        class="selected-city"
        v-if="citySelectionDivShown"
        close-text="Close Alert"
        dismissible
        >
        {{citySelectionText}}<strong>{{chosenCity}}</strong>
      </v-alert>
    </div>
  </v-container>
</template>

<script>
import accuweather from "./utils/accuweather";
import checkFlights from "./utils/checkFlights";
import CityInfoCard from './CityInfoCard'

export default {
  name: "WeatherAndFlightInfo",
  components:{ CityInfoCard },

  data: () => ({
    sourcecity: '',
    destinationcity: '',
    errors: '',
    errorLabel: 'Please correct the following error :',
    citySelectionDivShown: false,
    citySelectionText: 'You have selected to travel to : ',
    chosenCity: '',
    alert: true,
    submitButtonText: 'Check for an available flight ',
    cities: [
      {
        city: "Amsterdam",
        countryCode: "NL",
        airportCode: 'AMS',
        country: "Netherlands",
        weather: null,
        flights: []
      },

      {
        city: "Budapest",
        countryCode: "HU",
        airportCode: "BUD",
        country: "Hungary",
        weather: null,
        flights: []
      },
      {
        city: "Madrid",
        countryCode: "ES",
        airportCode: "MAD",
        country: "Spain",
        weather: null,
        flights: []
      }
    ]
  }),
  created: function() {
    this.fetchWeatherDetails();
  },
  methods: {
    fetchWeatherDetails() {
      this.cities.forEach(async (item, index) => {
        const res = await accuweather.getCityWeatherForcast(
          item.city,
          item.countryCode,
        );
        this.cities[index].weather = res;
      });
    },

    fetchFlightDetails() {
      if(!this.sourcecity) {
        this.errors = 'Source City is required.';
      } else if(this.errors){
        this.errors = '';
      }
      this.cities.forEach(async (item, index) => {
        if(item.airportCode !== this.sourcecity) {
            const flightsRes = await checkFlights.checkFlights(
            item.airportCode,
            this.sourcecity
          );
          this.cities[index].flights = flightsRes.data;
          this.cities[index].flights[0].currency = flightsRes.currency;
          this.cities[index]["error"] = false;
          this.sourcecity = '';
        } else {
          this.cities[index]["error"] = true;
        }
      });
    },
    chooseTheCity(chosenCity) {
      if(chosenCity) {
        this.chosenCity = chosenCity;
        this.citySelectionDivShown = true;
      }
    }
  }
};
</script>
<style>

.selected-city {
  justify-content: center;
  font-size: 25px;
}
</style>