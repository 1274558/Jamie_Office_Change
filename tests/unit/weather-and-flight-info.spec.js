import Vuetify from 'vuetify';
import Vue from "vue"
import WeatherAndFlightInfo from "@/components/WeatherAndFlightInfo";
import { createLocalVue, shallowMount } from "@vue/test-utils";

const localVue = createLocalVue()

describe("WeatherAndFlightInfo.vue", () => {
    let wrapper;
    let vuetify;

    beforeEach(() => {
        vuetify = new Vuetify();
        Vue.use(vuetify);
        wrapper = shallowMount(WeatherAndFlightInfo, {
            localVue,
            vuetify,
            propsData: {
                sourcecity: '',
                destinationcity: '',
                errors: '',
                chosenCity: '',
                alert: true,
            }
        });
    })

    it('renders', () => {
        wrapper.vm.sourcecity = '';
        expect(wrapper.exists()).toBe(true);
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('should throw an error when there is no value in an input field', () => {
        wrapper.vm.sourcecity = '';
        wrapper.vm.fetchFlightDetails();
        expect(wrapper.vm.errors).toBeTruthy();
    });
    it('should not throw an error when input field contains a value', () => {
        wrapper.vm.sourcecity = 'Test';
        wrapper.vm.fetchFlightDetails();
        expect(wrapper.vm.errors).toBeFalsy();
    });
})