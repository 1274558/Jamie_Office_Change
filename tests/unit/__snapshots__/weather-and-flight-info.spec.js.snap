// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherAndFlightInfo.vue renders 1`] = `
<v-container sourcecity="" destinationcity="" errors="" chosencity="" alert="true">
  <v-row>
    <v-col>
      <v-text-field label="Where are you flying from? (Enter airport code)" required=""></v-text-field>
      <!---->
      <v-btn type="submit" color="warning" dark="">
        Check for an available flight
      </v-btn>
    </v-col>
  </v-row>
  <v-row>
    <v-col cols="4">
      <city-info-card-stub citydata="[object Object]"></city-info-card-stub>
    </v-col>
    <v-col cols="4">
      <city-info-card-stub citydata="[object Object]"></city-info-card-stub>
    </v-col>
    <v-col cols="4">
      <city-info-card-stub citydata="[object Object]"></city-info-card-stub>
    </v-col>
  </v-row>
  <div>
    <!---->
  </div>
</v-container>
`;
